iroh-util/src/lock.rs

```diff
@@ -28,52 +35,143 @@ impl ProgramLock {
+    pub fn acquire(&mut self) -> AnyhowResult<()> {
+        // ensure path to lock exists
+        std::fs::create_dir_all(&crate::iroh_data_root()?)?;
+        let mut file = OpenOptions::new()
+            .write(true)
+            .create_new(true)
+            .open(&self.path)?;
+        file.write_all(process::id().to_string().as_bytes())?;
+        let file = Rc::new(file);
 
         file_guard::lock(file, Lock::Exclusive, 0, 1)
             .map(|lock| self.lock = Some(lock))
             .map_err(|err| err.into())
     }
 }
 
+/// Attempt to remove a stray lock file that wasn't cleaned up, returns true
+/// if a lock is successfully deleted, and will only attempt to delete if the
+/// lock is not currently held
+pub fn try_cleanup_dead_lock(prog_name: &str) -> AnyhowResult<bool> {
+    let lock = ProgramLock {
+        path: crate::iroh_data_path(&format!("{}.lock", prog_name))?,
+        lock: None,
+    };
+    if lock.is_locked() {
+        info!("lock {} is currently active, cannot remove", prog_name);
+        return Ok(false);
+    }
+    match std::fs::remove_file(lock.path) {
+        Err(e) => {
+            info!("error removing {} lockfile: {}", prog_name, e);
+            Err(anyhow!("removing dead lockfile: {}", e))
+        }
+        Ok(_) => {
+            info!("removed dead {} lockfile", prog_name);
+            Ok(true)
+        }
+    }
+}
+
+/// Report Process ID stored in a lock file
+pub fn read_lock_pid(prog_name: &str) -> Result<u32, LockError> {
+    let path = crate::iroh_data_path(&format!("{}.lock", prog_name))
+        .map_err(|e| LockError::Uncategorized(e.to_string()))?;
+    read_lock(path)
+}
+
+fn read_lock(path: PathBuf) -> Result<u32, LockError> {
+    let mut file = File::open(&path).map_err(|e| match e.kind() {
+        ErrorKind::NotFound => LockError::NoLock(path.clone()),
+        e => LockError::Uncategorized(e.to_string()),
+    })?;
+
+    let mut pid = String::new();
+    file.read_to_string(&mut pid)
+        .map_err(|_| LockError::CorruptLock(path.clone()))?;
+    let pid = pid
+        .parse::<u32>()
+        .map_err(|_| LockError::CorruptLock(path.clone()))?;
+    Ok(pid)
+}
+
+/// LockError classifies non-generic errors related to program locks
+#[derive(Error, Debug)]
+pub enum LockError {
+    #[error("No lock file at {0}")]
+    NoLock(PathBuf),
+    /// Failure to parse contents of lock file
+    #[error("Corrupt lock file contents at {0}")]
+    CorruptLock(PathBuf),
+    /// catchall error type
+    #[error("{0}")]
+    Uncategorized(String),
+}
+
 #[cfg(all(test, unix))]
 mod test {
     use super::*;
 
     fn create_test_lock(name: &str) -> ProgramLock {
         ProgramLock {
-            path: PathBuf::new().join(name),
+            path: PathBuf::from(name),
             lock: None,
         }
     }
 
+    #[test]
+    fn test_corrupt_lock() {
+        let path = PathBuf::from("lock.lock");
+        let mut f = File::create(&path).unwrap();
```

Comment:
I'm confused where this stores the file it creates. Does this get created in cwd of the test runner?

I've used the `tempdir` crate https://crates.io/crates/tempdir to ensure I have a safe temporary directory for writing files during testing that gets cleaned up after the test completes.

I see that this pattern is already established in other tests so I guess it hasn't hurt things yet.

Reply:
`tempdir` is deprecated, but please use `tempfile::Tempdir` for all testing directory uses